### PR TITLE
Feature/model aware dpg

### LINKED ISSUE #41: Build ops for computing MLE model learning loss

### LINKED ISSUE #42: Build ops for optimizing the dynamics model parameters via MLE

### PR DESCRIPTION
**Pull request by 0xangelo**
This PR introduces basic model training and model-aware DPG in off-policy MAPO. Hyperparameters are not tuned yet and model regularization is not used. There are still other types of model-aware losses to be implemented.

Closes #42 
Closes #41

### CODE REVIEW

File: mapo/agents/mapo/mapo_model.py [hunk 1]
```diff
@@ -49,21 +87,71 @@ def forward(self, input_dict, state, seq_lens):
...
+        """List of (main, target) variable pairs for applicable models."""
+        return [
+            (main, target)
+            for key in self.target_models
+            for main, target in zip(
+                self.models[key].variables, self.target_models[key].variables
+            )
+        ]
+
+    def get_actions(self, obs, target=False):
```
Comment (thiagopbueno) [inline mapo/agents/mapo/mapo_model.py]: I would call this method and the subsequent ones `compute_XXX` to make clear this is not a simple getter method. It actually adds several computation nodes to the graph.

Contrast it to the `_get_model` method which really is a getter method that simply return from a dict...

File: mapo/agents/mapo/off_mapo_policy.py [hunk 2]
```diff
@@ -71,20 +80,34 @@ def _build_critic_loss(policy, batch_tensors):
...
-    policy.loss_stats["qpi_mean"] = tf.reduce_mean(policy_action_value)
-    return -tf.reduce_mean(policy_action_value)
+    gamma = policy.config["gamma"]
+    model = policy.model
+    n_samples = policy.config["branching_factor"]
+
+    policy_action = model.get_actions(obs)
+    next_state_dist = model.next_state_dist(obs, policy_action)
+    sampled_next_state = next_state_dist.sample((n_samples,))
+    next_state_log_prob = tf.reduce_sum(
```
Comment (thiagopbueno) [inline mapo/agents/mapo/off_mapo_policy.py]: Both `sample()` and `log_prob()` methods are already implemented in the `GaussianDynamicsModel` class.

Why have you re-implemented them here?

Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: BTW... I will define a base class for Dynamics models so to make this interface standard...

File: mapo/agents/mapo/off_mapo_policy.py [hunk 3]
```diff
@@ -10,15 +10,24 @@
 from mapo.agents.mapo.mapo_policy import MAPOTFPolicy
 
 
+def _build_dynamics_loss(policy, batch_tensors):
```
Comment (thiagopbueno) [inline mapo/agents/mapo/off_mapo_policy.py]: I guess this function would be better called `_build_dynamics_mle_loss()` ...

File: mapo/agents/mapo/off_mapo_policy.py (hunk 2 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: Another thing... for the same reason you stopped the gradient on `sample_next_state` when computing the `next_state_value`, you should also stop the gradient when computing `next_state_log_prob` ... otherwise the optimizer would backpropagate through the sampling process as well in addition to the backpropagation through the distribution parameters in the log_prob computation...

File: mapo/agents/mapo/off_mapo_policy.py [hunk 4]
```diff
@@ -71,20 +80,34 @@ def _build_critic_loss(policy, batch_tensors):
...
+    )
+    return model_aware_policy_loss
 
 
-def build_actor_critic_losses(policy, batch_tensors):
-    """Contruct actor (DPG) and critic (Fitted Q) losses."""
+def build_mapo_losses(policy, batch_tensors):
+    """Contruct actor (MADPG), critic (Fitted Q) and dynamics (MLE) losses."""
     policy.loss_stats = {}
+    policy.dynamics_loss = _build_dynamics_loss(policy, batch_tensors)
```
Comment (thiagopbueno) [inline mapo/agents/mapo/off_mapo_policy.py]: What is the easiest way to have access to the config dict so to choose the appropriate _build_dynamics_loss (between MLE and PG-aware) ?

File: mapo/agents/mapo/mapo_model.py [hunk 5]
```diff
@@ -6,34 +6,72 @@
...
+                target_models["twin_q_net"] = build_continuous_q_function(
+                    obs_space, action_space, model_config["custom_options"]["critic"]
+                )
+            self.target_models = target_models
+            self.register_variables(
+                [var for model in target_models.values() for var in model.variables]
+            )
 
     @override(TFModelV2)
     def forward(self, input_dict, state, seq_lens):
```
Comment (thiagopbueno) [inline mapo/agents/mapo/mapo_model.py:77]: Just as of curiosity, why not to use `forward()` to compute the `policy_action`, `sampled_next_state`, and `next_state_value`...?

This way you could simplify `_build_actor_loss(policy, batch_tensors)` in `off_mapo_policy.py` module..

Comment (thiagopbueno) [reply mapo/agents/mapo/mapo_model.py:77]: As a side effect, it would become easier to test these computation that are currently buried in the `_build_actor_loss` function...

File: mapo/agents/mapo/off_mapo_policy.py [hunk 6]
```diff
@@ -302,26 +329,24 @@ class TargetUpdatesMixin:  # pylint: disable=too-few-public-methods
...
         update_target_expr = [
-            target_var.assign(tau * var + (1.0 - tau) * target_var)
-            for var, target_var in zip(main_variables, target_variables)
+            target.assign(tau * main + (1.0 - tau) * target)
+            for main, target in self.model.main_and_target_variables
         ]
         return tf.group(*update_target_expr)
 
 
 OffMAPOTFPolicy = MAPOTFPolicy.with_updates(  # pylint: disable=invalid-name
```
Comment (thiagopbueno) [inline mapo/agents/mapo/off_mapo_policy.py]: What does `OffMAPOTFPolicy` depend / extend on `MAPOTFPolicy` ?

File: mapo/agents/mapo/tests/test_exploration_state.py [hunk 7]
```diff
@@ -22,6 +23,7 @@ def test_deterministic_evaluation():
     assert np.allclose(action1, action2)
 
 
+@pytest.mark.skip
```
Comment (thiagopbueno) [inline mapo/agents/mapo/tests/test_exploration_state.py]: Why?

File: mapo/agents/mapo/tests/test_exploration_state.py [hunk 8]
```diff
@@ -42,6 +44,7 @@ def rvs(size=1):
     assert p_value >= 0.05
 
 
+@pytest.mark.skip
```
Comment (thiagopbueno) [inline mapo/agents/mapo/tests/test_exploration_state.py]: Why?

**Review by thiagopbueno — CHANGES_REQUESTED** [review]

File: mapo/agents/mapo/off_mapo_policy.py (hunk 4 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: I guess we can just call `policy.config` here, right?

File: mapo/agents/mapo/mapo_model.py (hunk 1 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/mapo_model.py]: I assumed these `get_XXX` methods aren't overriding base methods... but if they are it would be better to decorate them with `@override()` ... and in that case, forget about my previous comments...

Comment (0xangelo) [reply mapo/agents/mapo/mapo_model.py]: You're right, it makes sense to make the computation aspect explicit. I'll make sure to update this.

File: mapo/agents/mapo/off_mapo_policy.py (hunk 2 above)
Comment (0xangelo) [reply mapo/agents/mapo/off_mapo_policy.py]: > Both `sample()` and `log_prob()` methods are already implemented in the `GaussianDynamicsModel` class.
> 
> Why have you re-implemented them here?

I didn't want to use a separate forward pass through the dynamics just to compute the `log_prob`. Maybe TensorFlow takes care of optimizing this, but from my experience using PyTorch, I find it better to use the same `dist` to sample the actions and compute the `log_prob`.

Comment (0xangelo) [reply mapo/agents/mapo/off_mapo_policy.py]: > Another thing... for the same reason you stopped the gradient on `sample_next_state` when computing the `next_state_value`, you should also stop the gradient when computing `next_state_log_prob` ... otherwise the optimizer would backpropagate through the sampling process as well in addition to the backpropagation through the distribution parameters in the log_prob computation...

Probably another effect of using PyTorch before. There, all tensors created by `sample()` from a `Distribution` have no `grad_fn`. Reparameterizable distributions have an explicit `rsample()` method  which propagates gradients.

Since I'm not sure if the same applies here, I'll call `stop_gradient` on the sample as suggested. Thanks!

File: mapo/agents/mapo/off_mapo_policy.py (hunk 3 above)
Comment (0xangelo) [reply mapo/agents/mapo/off_mapo_policy.py]: Good point, we can switch between different losses this way.

File: mapo/agents/mapo/off_mapo_policy.py (hunk 4 above)
Comment (0xangelo) [reply mapo/agents/mapo/off_mapo_policy.py]: That's right. I didn't consider putting it here since the gradient-aware loss wasn't implemented, but I'll do it now.

File: mapo/agents/mapo/off_mapo_policy.py (hunk 2 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: You are right it would be better not to add another forward call through the dynamics... However I think it's best to implement a `log_prob_sampled` method in the dynamics model to handle this    implementation and also to take care of the stop_gradient issue...

I will do that right away.

File: mapo/agents/mapo/mapo_model.py (hunk 5 above)
Comment (0xangelo) [reply mapo/agents/mapo/mapo_model.py:77]: Unfortunately `ModelV2`'s `__call__` method requires us to return a single tensor of shape `(None, output_size)`. That's why there's the hack in `__init__` where we call `super().__init__` with a predefined output size, and then return in `forward` the flattened observation. I had to do this because `self.model` is called internally by DynamicTFPolicy (see [here](https://hackmd.io/@angelovt/By-MYD77r#build-model-output)).

IMO, this is a bad design decision by the developers, as I would also love to do something like:
```python
policy_action, sampled_next_state, next_state_value = self.model(inputs)
``` 
The only workaround I found was to separate the outputs in different methods, as I didn't want to group everything in one tensor and split it afterwards. We can discuss this if you think it's a good idea though.

File: mapo/agents/mapo/off_mapo_policy.py (hunk 6 above)
Comment (0xangelo) [reply mapo/agents/mapo/off_mapo_policy.py]: They share two methods, which are 
```python
    optimizer_fn=create_separate_optimizers,
    gradients_fn=compute_separate_gradients,
```
I can copy the two from `MAPOTFPolicy` and declare `OffMAPOTFPolicy` with `build_tf_policy` to avoid confusion.

File: mapo/agents/mapo/tests/test_exploration_state.py (hunk 7 above)
Comment (0xangelo) [reply mapo/agents/mapo/tests/test_exploration_state.py]: I think I explained it in this commit bc5f641

File: mapo/agents/mapo/tests/test_exploration_state.py (hunk 8 above)
Comment (0xangelo) [reply mapo/agents/mapo/tests/test_exploration_state.py]: Same as above.

File: mapo/agents/mapo/off_mapo_policy.py (hunk 2 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: I've just pushed a commit in feature/unimodal-gaussian-model with a log_prob_sampled that you can use... let me know what you think

File: mapo/agents/mapo/tests/test_exploration_state.py (hunk 7 above)
Comment (0xangelo) [reply mapo/agents/mapo/tests/test_exploration_state.py]: I could maybe try to increase the batch size.

File: mapo/agents/mapo/mapo_model.py (hunk 5 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/mapo_model.py:77]: Maybe a simpler solution would be to compute everything in forward, but only return the policy_action... the other tensors you could plug as attributes of the model ...

just as in the simple example in the documentation below in which the value function tensor is set to `self._value_out` ..

https://ray.readthedocs.io/en/latest/rllib-models.html?highlight=TFModelV2#ray.rllib.models.tf.tf_modelv2.TFModelV2.forward

What do you think of this solution? Would it be worth it so to make the code more testable?

File: mapo/agents/mapo/tests/test_exploration_state.py (hunk 7 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/tests/test_exploration_state.py]: Indeed, you could try to increase the batch size and if this slows down the test too much consider decorate it with `pytest.mark.slow` ...

Comment (0xangelo) [reply mapo/agents/mapo/tests/test_exploration_state.py]: Actually, I think I made a grave error... I'm testing if the output distribution of the policy is a gaussian with the mean in the deterministic action and standard deviation defined by the config. The problem is that I clip the actions in `build_action_sampler` to the desired action space. No wonder the test fails, even when (and specially when) I increase the batch size. 

I'll make sure to use an unbounded action space for this test.

File: mapo/agents/mapo/mapo_model.py (hunk 5 above)
Comment (0xangelo) [reply mapo/agents/mapo/mapo_model.py:77]: I find very hacky since it creates side-effects (changes the model's attributes) after every forward pass, which I'm not a fan of. I think we can still test the loss computation using the custom methods I created. 

Shall we revisit this issue later? (if the need arises)

Comment (thiagopbueno) [reply mapo/agents/mapo/mapo_model.py:77]: sure! very tangential issue here... nothing to be done right now! thanks

File: mapo/agents/mapo/off_mapo_policy.py (hunk 6 above)
Comment (thiagopbueno) [reply mapo/agents/mapo/off_mapo_policy.py]: I personally would prefer to avoid implicit dependencies like this in the future... but no worries

**Review by thiagopbueno — APPROVED** [review]
